src/shared/bootspec.c

```diff
@@ -950,9 +943,11 @@ static int boot_entries_find_unified_addons(
                 if (!location)
                         return log_oom();
 
-                r = insert_boot_entry_addon(addons, TAKE_PTR(location), TAKE_PTR(cmdline));
+                r = insert_boot_entry_addon(addons, location, cmdline);
                 if (r < 0) {
-                        free(addons);
+                        free(addons->items);
+                        free(location);
+                        free(cmdline);
                         return r;
                 }
```

Comment:
usually, how we do this is to handle this in the success path, not in the error path.

or in other words: in the success path (i.e. right here after the `}`) add `TAKE_PTR(location);` and `TAKE_PTR(cmdline);`. (without assigning them to anything. this will just set the pointers to NULL, i.e. is equivalent to `location = NULL;` and `cmdline = NULL`, but it's more explanatory we think since it makes clear that ownership of the pointer was passed away

Reply:
i.e.:

```c
r = insert_boot_entry_addon(addons, location, cmdline);
if (r < 0)
        return r;

TAKE_PTR(location);
TAKE_PTR(cmdline);
```